File: src/formatters/terminal.js
const c = {
  reset: '\x1b[0m',
  bold: '\x1b[1m',
  dim: '\x1b[2m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
  white: '\x1b[37m',
  bgBlue: '\x1b[44m',
};

function header(emoji, title) {
  return `\n${c.bold}${c.cyan}${emoji}  ${title}${c.reset}\n${'─'.repeat(50)}`;
}

function formatWeather(data) {
  if (data.error) return `${header('🌤️', 'Weather')}  \n  ${c.red}Error: ${data.error}${c.reset}`;
  const lines = [
    header('🌤️', `Weather — ${data.location}`),
    `  ${c.bold}${data.condition}${c.reset}  ${data.temp_f}°F (feels like ${data.feels_like_f}°F)`,
    `  ${c.dim}High: ${data.high_f}°F  Low: ${data.low_f}°F${c.reset}`,
    `  💧 Humidity: ${data.humidity}%  🌬️  Wind: ${data.wind_mph} mph ${data.wind_dir}`,
    `  ☀️  UV: ${data.uv_index}  🌧️  Rain: ${data.precip_chance}%`,
    `  🌅 Sunrise: ${data.sunrise}  🌇 Sunset: ${data.sunset}`,
  ];
  return lines.join('\n');
}

function formatCalendar(data) {
  if (data.error) return `${header('📅', 'Calendar')}\n  ${c.red}Error: ${data.error}${c.reset}`;
  if (!data.available) return `${header('📅', 'Calendar')}\n  ${c.dim}${data.note}${c.reset}`;
  if (!data.events || data.events.length === 0) {
    return `${header('📅', 'Calendar')}\n  ${c.green}No events today — wide open!${c.reset}`;
  }
  const lines = [header('📅', `Calendar — ${data.count} event${data.count !== 1 ? 's' : ''}`)];
  for (const e of data.events) {
    if (e.raw) {
      lines.push(`  ${e.raw}`);
    } else {
      const loc = e.location ? `  ${c.dim}@ ${e.location}${c.reset}` : '';
      lines.push(`  ${c.bold}${e.start_time}${c.reset} ${e.title}${loc}`);
    }
  }
  return lines.join('\n');
}

function formatGit(data) {
  if (data.error) return `${header('📦', 'Git Status')}\n  ${c.red}Error: ${data.error}${c.reset}`;
  const lines = [header('📦', `Git Status — ${data.totalRepos} repos`)];

  if (data.dirtyRepos > 0) {
    lines.push(`  ${c.yellow}⚠ ${data.dirtyRepos} repo${data.dirtyRepos > 1 ? 's' : ''} with uncommitted changes${c.reset}`);
  }
  if (data.reposWithRecentCommits > 0) {
    lines.push(`  ${c.green}✓ ${data.reposWithRecentCommits} repo${data.reposWithRecentCommits > 1 ? 's' : ''} with commits in last 24h${c.reset}`);
  }

  const interesting = data.repos.filter(r => r.uncommitted > 0 || r.ahead > 0 || r.behind > 0 || r.recentCommits > 0);
  if (interesting.length > 0) {
    lines.push('');
    for (const r of interesting) {
      const flags = [];
      if (r.uncommitted > 0) flags.push(`${c.yellow}${r.uncommitted} uncommitted${c.reset}`);
      if (r.ahead > 0) flags.push(`${c.green}↑${r.ahead}${c.reset}`);
      if (r.behind > 0) flags.push(`${c.red}↓${r.behind}${c.reset}`);
      if (r.recentCommits > 0) flags.push(`${c.cyan}${r.recentCommits} recent${c.reset}`);
      lines.push(`  ${c.bold}${r.name}${c.reset} (${r.branch}) — ${flags.join(', ')}`);
    }
  }

  if (data.errors) {
    for (const e of data.errors) lines.push(`  ${c.red}${e}${c.reset}`);
  }

  if (interesting.length === 0 && data.dirtyRepos === 0) {
    lines.push(`  ${c.green}All clean!${c.reset}`);
  }

  return lines.join('\n');
}

function formatSystem(data) {
  if (data.error) return `${header('🖥️', 'System')}\n  ${c.red}Error: ${data.error}${c.reset}`;
  const memColor = parseInt(data.memory.percent) > 80 ? c.red : parseInt(data.memory.percent) > 60 ? c.yellow : c.green;
  const lines = [
    header('🖥️', 'System Health'),
    `  ⏱️  Uptime: ${data.uptime}  |  CPUs: ${data.cpus}`,
    `  📊 Load: ${data.load['1m']} / ${data.load['5m']} / ${data.load['15m']}`,
    `  🧠 Memory: ${memColor}${data.memory.percent}%${c.reset} (${data.memory.used_gb}/${data.memory.total_gb} GB)`,
  ];

  if (data.disks) {
    lines.push(`  💾 Disks:`);
    for (const d of data.disks) {
      const dColor = d.warning ? c.red : c.green;
      const warn = d.warning ? ' ⚠️' : '';
      lines.push(`     ${d.mount}: ${dColor}${d.percent}%${c.reset} (${d.used}/${d.size})${warn}`);
    }
  }

  return lines.join('\n');
}

function formatK8s(data) {
  if (data.error) return `${header('☸️', 'Kubernetes')}\n  ${c.red}Error: ${data.error}${c.reset}`;
  if (!data.available) return `${header('☸️', 'Kubernetes')}\n  ${c.dim}${data.note}${c.reset}`;

  const lines = [header('☸️', 'Kubernetes')];

  if (data.nodes) {
    const readyStr = data.nodesReady === data.nodeCount
      ? `${c.green}${data.nodesReady}/${data.nodeCount} ready${c.reset}`
      : `${c.red}${data.nodesReady}/${data.nodeCount} ready${c.reset}`;
    lines.push(`  🖧  Nodes: ${readyStr}`);
  }

  lines.push(`  📦 Pods: ${data.totalPods} total`);

  if (data.unhealthyCount > 0) {
    lines.push(`  ${c.red}⚠ ${data.unhealthyCount} unhealthy pod${data.unhealthyCount > 1 ? 's' : ''}:${c.reset}`);
    for (const p of data.unhealthyPods.slice(0, 10)) {
      lines.push(`     ${c.red}${p.namespace}/${p.name} (${p.phase})${c.reset}`);
    }
  } else {
    lines.push(`  ${c.green}✓ All pods healthy${c.reset}`);
  }

  if (data.restartIssueCount > 0) {
    lines.push(`  ${c.yellow}🔄 ${data.restartIssueCount} container${data.restartIssueCount > 1 ? 's' : ''} with restart loops:${c.reset}`);
    for (const r of data.restartIssues.slice(0, 10)) {
      lines.push(`     ${c.yellow}${r.namespace}/${r.pod}:${r.container} (${r.restarts} restarts)${c.reset}`);
    }
  }

  return lines.join('\n');
}

function formatTerminal(results) {
  const now = new Date();
  const dateStr = now.toLocaleDateString('en-US', { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' });
  const timeStr = now.toLocaleTimeString('en-US', { hour: '2-digit', minute: '2-digit' });

  const parts = [
    `\n${c.bold}${c.bgBlue}${c.white} ☀️  DAILY BRIEFING — ${dateStr} ${timeStr} ${c.reset}\n`,
  ];

  const s = results.sections;
  if (s.weather) parts.push(formatWeather(s.weather));
  if (s.calendar) parts.push(formatCalendar(s.calendar));
  if (s.git) parts.push(formatGit(s.git));
  if (s.system) parts.push(formatSystem(s.system));
  if (s.kubernetes) parts.push(formatK8s(s.kubernetes));

  parts.push(`\n${c.dim}Generated at ${results.timestamp}${c.reset}\n`);

  return parts.join('\n');
}

module.exports = { formatTerminal };

File: daily_briefing/formatters/terminal.py
"""Terminal formatter — colorful ANSI output."""

from datetime import datetime

# ANSI codes
RESET = "\033[0m"
BOLD = "\033[1m"
DIM = "\033[2m"
RED = "\033[31m"
GREEN = "\033[32m"
YELLOW = "\033[33m"
BLUE = "\033[34m"
CYAN = "\033[36m"
WHITE = "\033[37m"
BG_BLUE = "\033[44m"


def _header(emoji, title):
    return f"\n{BOLD}{CYAN}{emoji}  {title}{RESET}\n{'─' * 50}"


def _fmt_weather(data):
    if data.get("error"):
        return f"{_header('🌤️', 'Weather')}\n  {RED}Error: {data['error']}{RESET}"
    return "\n".join([
        _header("🌤️", f"Weather — {data['location']}"),
        f"  {BOLD}{data['condition']}{RESET}  {data['temp_f']}°F (feels like {data['feels_like_f']}°F)",
        f"  {DIM}High: {data['high_f']}°F  Low: {data['low_f']}°F{RESET}",
        f"  💧 Humidity: {data['humidity']}%  🌬️  Wind: {data['wind_mph']} mph {data['wind_dir']}",
        f"  ☀️  UV: {data['uv_index']}  🌧️  Rain: {data['precip_chance']}%",
        f"  🌅 Sunrise: {data['sunrise']}  🌇 Sunset: {data['sunset']}",
    ])


def _fmt_calendar(data):
    if data.get("error"):
        return f"{_header('📅', 'Calendar')}\n  {RED}Error: {data['error']}{RESET}"
    if not data.get("available"):
        return f"{_header('📅', 'Calendar')}\n  {DIM}{data.get('note', 'Not configured')}{RESET}"
    events = data.get("events", [])
    if not events:
        return f"{_header('📅', 'Calendar')}\n  {GREEN}No events today — wide open!{RESET}"
    count = data.get("count", len(events))
    lines = [_header("📅", f"Calendar — {count} event{'s' if count != 1 else ''}")]
    for e in events:
        loc = f"  {DIM}@ {e['location']}{RESET}" if e.get("location") else ""
        lines.append(f"  {BOLD}{e['start_time']}{RESET} {e['title']}{loc}")
    return "\n".join(lines)


def _fmt_git(data):
    if data.get("error"):
        return f"{_header('📦', 'Git Status')}\n  {RED}Error: {data['error']}{RESET}"
    lines = [_header("📦", f"Git Status — {data['totalRepos']} repos")]

    if data["dirtyRepos"] > 0:
        n = data["dirtyRepos"]
        lines.append(f"  {YELLOW}⚠ {n} repo{'s' if n > 1 else ''} with uncommitted changes{RESET}")
    if data["reposWithRecentCommits"] > 0:
        n = data["reposWithRecentCommits"]
        lines.append(f"  {GREEN}✓ {n} repo{'s' if n > 1 else ''} with commits in last 24h{RESET}")

    interesting = [
        r for r in data["repos"]
        if r["uncommitted"] > 0 or r.get("ahead", 0) > 0 or r.get("behind", 0) > 0 or r["recentCommits"] > 0
    ]
    if interesting:
        lines.append("")
        for r in interesting:
            flags = []
            if r["uncommitted"] > 0:
                flags.append(f"{YELLOW}{r['uncommitted']} uncommitted{RESET}")
            if r.get("ahead", 0) > 0:
                flags.append(f"{GREEN}↑{r['ahead']}{RESET}")
            if r.get("behind", 0) > 0:
                flags.append(f"{RED}↓{r['behind']}{RESET}")
            if r["recentCommits"] > 0:
                flags.append(f"{CYAN}{r['recentCommits']} recent{RESET}")
            lines.append(f"  {BOLD}{r['name']}{RESET} ({r['branch']}) — {', '.join(flags)}")

    if data.get("errors"):
        for e in data["errors"]:
            lines.append(f"  {RED}{e}{RESET}")

    if not interesting and data["dirtyRepos"] == 0:
        lines.append(f"  {GREEN}All clean!{RESET}")

    return "\n".join(lines)


def _fmt_system(data):
    if data.get("error"):
        return f"{_header('🖥️', 'System')}\n  {RED}Error: {data['error']}{RESET}"

    mem = data.get("memory", {})
    pct = int(mem.get("percent", 0))
    mem_color = RED if pct > 80 else YELLOW if pct > 60 else GREEN

    lines = [
        _header("🖥️", "System Health"),
        f"  ⏱️  Uptime: {data['uptime']}  |  CPUs: {data['cpus']}",
        f"  📊 Load: {data['load']['1m']} / {data['load']['5m']} / {data['load']['15m']}",
        f"  🧠 Memory: {mem_color}{mem['percent']}%{RESET} ({mem['used_gb']}/{mem['total_gb']} GB)",
    ]

    if data.get("disks"):
        lines.append("  💾 Disks:")
        for d in data["disks"]:
            d_color = RED if d["warning"] else GREEN
            warn = " ⚠️" if d["warning"] else ""
            lines.append(f"     {d['mount']}: {d_color}{d['percent']}%{RESET} ({d['used']}/{d['size']}){warn}")

    return "\n".join(lines)


def _fmt_k8s(data):
    if data.get("error"):
        return f"{_header('☸️', 'Kubernetes')}\n  {RED}Error: {data['error']}{RESET}"
    if not data.get("available"):
        return f"{_header('☸️', 'Kubernetes')}\n  {DIM}{data.get('note', 'Not available')}{RESET}"

    lines = [_header("☸️", "Kubernetes")]

    if data.get("nodes"):
        ready = data["nodesReady"]
        total = data["nodeCount"]
        color = GREEN if ready == total else RED
        lines.append(f"  🖧  Nodes: {color}{ready}/{total} ready{RESET}")

    lines.append(f"  📦 Pods: {data['totalPods']} total")

    if data.get("unhealthyCount", 0) > 0:
        n = data["unhealthyCount"]
        lines.append(f"  {RED}⚠ {n} unhealthy pod{'s' if n > 1 else ''}:{RESET}")
        for p in data["unhealthyPods"][:10]:
            lines.append(f"     {RED}{p['namespace']}/{p['name']} ({p['phase']}){RESET}")
    else:
        lines.append(f"  {GREEN}✓ All pods healthy{RESET}")

    if data.get("restartIssueCount", 0) > 0:
        n = data["restartIssueCount"]
        lines.append(f"  {YELLOW}🔄 {n} container{'s' if n > 1 else ''} with restart loops:{RESET}")
        for r in data["restartIssues"][:10]:
            lines.append(f"     {YELLOW}{r['namespace']}/{r['pod']}:{r['container']} ({r['restarts']} restarts){RESET}")

    return "\n".join(lines)


def format_terminal(results):
    now = datetime.now()
    date_str = now.strftime("%A, %B %d, %Y")
    time_str = now.strftime("%I:%M %p").lstrip("0")

    parts = [
        f"\n{BOLD}{BG_BLUE}{WHITE} ☀️  DAILY BRIEFING — {date_str} {time_str} {RESET}\n",
    ]

    s = results["sections"]
    for key in ["weather", "calendar", "git", "system", "kubernetes"]:
        if key in s:
            formatter = {
                "weather": _fmt_weather,
                "calendar": _fmt_calendar,
                "git": _fmt_git,
                "system": _fmt_system,
                "kubernetes": _fmt_k8s,
            }[key]
            parts.append(formatter(s[key]))

    parts.append(f"\n{DIM}Generated at {results['timestamp']}{RESET}\n")
    return "\n".join(parts)
